### PR TITLE
Convert STM32 drivers to DT_INST

### PR DESCRIPTION
**Pull request by erwango**
Superseeds #22807

- [x] drivers/adc/adc_stm32.c
- [ ] drivers/can/can_stm32.c (cf below comment)
- [x] drivers/counter/counter_ll_stm32_rtc.c
- [ ] drivers/ethernet/eth_stm32_hal.c (Only impact `DT_DTCM_BASE_ADDRESS`. To be removed?)
- [x] drivers/flash/flash_stm32.c
- [ ] drivers/flash/flash_stm32f3x.c (Only impact `DT_FLASH_SIZE`. To be removed ?)
- [ ] drivers/flash/flash_stm32l4x.c (Only impact `DT_FLASH_SIZE`. be removed ?)
- [ ] drivers/gpio/gpio_stm32.c (cf below comment)
- [ ] drivers/i2c/i2c_ll_stm32.c (cf below comment)
- [ ] drivers/i2s/i2s_ll_stm32.c (cf below comment)
- [x] drivers/pwm/pwm_stm32.c
- [x] drivers/serial/uart_stm32.c
- [x] drivers/spi/spi_ll_stm32.c
- [ ] drivers/timer/stm32_lptim_timer.c (To be converted to DT first)
- [x] drivers/usb/device/usb_dc_stm32.c

Conversion of some drivers to DT_INST not possible yet, cf https://github.com/zephyrproject-rtos/zephyr/pull/22807#issuecomment-592481101

### CODE REVIEW

**Comment by erwango** [comment]
Following test is failed:
`902/919 qemu_xtensa               tests/kernel/sched/schedule_api/kernel.scheduler.multiq FAILED Failed (qemu 6.259s)`

File: drivers/spi/spi_ll_stm32.c [hunk 1]
```diff
@@ -533,33 +534,34 @@ static struct spi_stm32_data spi_stm32_dev_data_##id = {		\
...
 		    &spi_stm32_dev_data_##id, &spi_stm32_cfg_##id,	\
 		    POST_KERNEL, CONFIG_SPI_INIT_PRIORITY,		\
 		    &api_funcs);					\
 									\
 STM32_SPI_IRQ_HANDLER(id)
 
-#ifdef CONFIG_SPI_1
+#ifdef DT_INST_0_ST_STM32_SPI
+STM32_SPI_INIT(0)
+#endif
```
Comment (arnopo) [inline drivers/spi/spi_ll_stm32.c]: only a suggestion...you could add here and below /* DT_INST_x_ST_STM32_SPI */ comment to be coherent with other commits

File: drivers/usb/device/usb_dc_stm32.c [hunk 2]
```diff
@@ -55,10 +55,42 @@
 #include <logging/log.h>
 LOG_MODULE_REGISTER(usb_dc_stm32);
 
-#if defined(DT_USB_BASE_ADDRESS) && defined(DT_USB_HS_BASE_ADDRESS)
+#if defined(DT_COMPAT_ST_STM32_OTGFS) && defined(DT_COMPAT_ST_STM32_OTGHS)
```
Comment (arnopo) [inline drivers/usb/device/usb_dc_stm32.c]: DT_COMPAT_ST_STM32_USB seems also exclusive

**Review by arnopo — COMMENTED** [review]
Hi @erwango,
minor remarks else LGTM

File: drivers/adc/adc_stm32.c [hunk 3]
```diff
@@ -648,18 +648,20 @@ static struct adc_stm32_data adc_stm32_data_##index = {			\
...
-	irq_enable(DT_ADC_##index##_IRQ);				\
+	irq_enable(DT_INST_##index##_ST_STM32_ADC_IRQ_0);		\
 }
 
-#ifdef CONFIG_ADC_1
-STM32_ADC_INIT(1)
-#endif /* CONFIG_ADC_1 */
+#ifdef DT_INST_0_ST_STM32_ADC
+STM32_ADC_INIT(0);
+#endif /* DT_INST_0_ST_STM32_DMA */
```
[the remark this answers is not in the document]
Comment (erwango) [reply drivers/adc/adc_stm32.c]: Fixed, txs

File: drivers/spi/spi_ll_stm32.c (hunk 1 above)
Comment (erwango) [reply drivers/spi/spi_ll_stm32.c]: Done

File: drivers/usb/device/usb_dc_stm32.c (hunk 2 above)
Comment (erwango) [reply drivers/usb/device/usb_dc_stm32.c]: Indeed, but this error was made specifically for one series in which both versions are available in device tree description and where we should take care they are not activated at the same time.
This is not possible for other compatible ("st, stm32-usb") as its presence is already exclusive in dts files.

**Comment by erwango** [comment]
@arnopo, thanks for the review. Points fixed or commented.

**Comment by erwango** [comment]
One failed test:
`919/919 qemu_cortex_r5            tests/kernel/context/kernel.common                 FAILED Failed (qemu 6.129s)`

**Review by arnopo — APPROVED** [review]

**Comment by erwango** [comment]
Failing test:
`Failures
 Sanitycheck / qemu_x86_64:tests/kernel/mem_protect/syscalls/kernel.memory_protection.syscalls / tests/kernel/mem_protect/syscalls/kernel.memory_protection.syscalls`

**Comment by galak** [comment]
@erwango would it make sense to merge this and have new PRs so we don't have CI growing to large?  I know we'll need a second pass conversion based on what we are doing in PR #23245.

**Comment by erwango** [comment]
> @erwango would it make sense to merge this and have new PRs 

@galak Definitely!

**Review by ydamigos — APPROVED** [review]

File: boards/arm/olimex_stm32_e407/pinmux.c [hunk 4]
```diff
@@ -40,16 +40,14 @@ static const struct pin_config pinconf[] = {
...
 	{STM32_PIN_PG14, STM32F4_PINMUX_FUNC_PG14_ETH},
 #endif /* CONFIG_ETH_STM32_HAL */
-#ifdef CONFIG_USB_DC_STM32
-#ifdef DT_USB_BASE_ADDRESS
+#ifdef DT_INST_0_ST_STM32_OTGFS
 	{STM32_PIN_PA11, STM32F4_PINMUX_FUNC_PA11_OTG_FS_DM},
 	{STM32_PIN_PA12, STM32F4_PINMUX_FUNC_PA12_OTG_FS_DP},
-#endif /* DT_USB_BASE_ADDRESS */
-#ifdef DT_USB_HS_BASE_ADDRESS
+#endif /* DT_COMPAT_ST_STM32_OTGFS */
```
Comment (galak) [inline boards/arm/olimex_stm32_e407/pinmux.c]: endif comment doesn't match ifdef

File: boards/arm/olimex_stm32_e407/pinmux.c [hunk 5]
```diff
@@ -40,16 +40,14 @@ static const struct pin_config pinconf[] = {
...
+#ifdef DT_INST_0_ST_STM32_OTGHS
 	{STM32_PIN_PB14, STM32F4_PINMUX_FUNC_PB14_OTG_HS_DM},
 	{STM32_PIN_PB15, STM32F4_PINMUX_FUNC_PB15_OTG_HS_DP},
-#endif /* DT_USB_HS_BASE_ADDRESS */
-#endif  /* CONFIG_USB_DC_STM32 */
+#endif /* DT_COMPAT_ST_STM32_OTGHS */
```
Comment (galak) [inline boards/arm/olimex_stm32_e407/pinmux.c]: endif comment doesn't match ifdef

File: boards/arm/olimex_stm32_e407/pinmux.c (hunk 4 above)
Comment (erwango) [reply boards/arm/olimex_stm32_e407/pinmux.c]: Done, txs

File: boards/arm/olimex_stm32_e407/pinmux.c (hunk 5 above)
Comment (erwango) [reply boards/arm/olimex_stm32_e407/pinmux.c]: Done, txs

**Review by galak — APPROVED** [review]

File: soc/arm/st_stm32/stm32f4/dts_fixup.h [hunk 6]
```diff
@@ -271,30 +271,6 @@
 #define DT_I2S_6_DMA_FEATURES_RX		\
 			DT_ST_STM32_I2S_40015400_RX_DMAS_FEATURES
 
-#define DT_FLASH_DEV_NAME		DT_INST_0_ST_STM32F4_FLASH_CONTROLLER_LABEL
```
Comment (mnkp) [inline soc/arm/st_stm32/stm32f4/dts_fixup.h]: The `DT_FLASH_DEV_NAME` was accidentally removed by "drivers/usb: usb_dc_stm32: Convert to DT_INST" commit a778bcd45a560d58db08da8673b360390c74095f. We need a follow up patch. Some applications, e.g. MCUBoot fail to compile.

File: soc/arm/st_stm32/stm32f0/dts_fixup.h [hunk 7]
```diff
@@ -119,8 +119,7 @@
 #define DT_CAN_1_CLOCK_BUS		DT_ST_STM32_CAN_40006400_CLOCK_BUS
 #define DT_CAN_1_CLOCK_BITS		DT_ST_STM32_CAN_40006400_CLOCK_BITS
 
-#define DT_FLASH_DEV_BASE_ADDRESS	DT_ST_STM32F0_FLASH_CONTROLLER_40022000_BASE_ADDRESS
-#define DT_FLASH_DEV_NAME		DT_ST_STM32F0_FLASH_CONTROLLER_40022000_LABEL
+#define DT_FLASH_DEV_NAME		DT_INST_0_ST_STM32F0_FLASH_CONTROLLER_LABEL
```
Comment (mnkp) [inline soc/arm/st_stm32/stm32f0/dts_fixup.h]: The `flash_stm32.c` driver is using `DT_INST_0_SOC_NV_FLASH_LABEL` as a flash name
```
DT_INST_0_SOC_NV_FLASH_LABEL="FLASH_STM32"
```
however the `DT_FLASH_DEV_NAME` is mapped to the flash controller name
```
DT_INST_0_ST_STM32F0_FLASH_CONTROLLER_LABEL="FLASH_CTRL"

```
The two don't match. This is causing run time error in MCUBoot.

Comment (erwango) [reply soc/arm/st_stm32/stm32f0/dts_fixup.h]: @mnkp, thanks for reporting. I'm fixing it.

**Pull request merged**